### PR TITLE
Add named prerequisites

### PR DESCRIPTION
**Pull request by prachigauriar**
- [x] Implement methods for getting/setting named prerequisites
- [x] Write in-code documentation
- [x] Update unit tests
- [x] Update readme
- [x] Update version info in files

### CODE REVIEW

File: Task/Tasks/TSKTask.h [hunk 1]
```diff
@@ -161,21 +161,51 @@ extern NSString *const TSKTaskDidStartNotification;
  */
 @property (nonatomic, weak, readonly) TSKWorkflow *workflow;
 
+/*! 
+ @abstract The required prerequisite keys that must be fulfilled for the task to run.
```
Comment (prachigauriar) [inline Task/Tasks/TSKTask.h]: Improve this.

File: Task/Tasks/TSKTask.h [hunk 2]
```diff
@@ -342,6 +378,29 @@ extern NSString *const TSKTaskDidStartNotification;
...
 /*!
+ @abstract Returns the results of all the receiver’s unkeyed prerequisite tasks in an array.
+ @discussion The returned array will contain NSNull elements for each task that returns nil. 
+     The order of the array is arbitrary.
+ @result An array containing the results of all the receiver’s unkeyed prerequisite tasks.
+ */
+- (NSArray *)allUnkeyedPrerequisiteResults;
+
+/*!
+ @abstract Returns the results of all receiver’s keyed prerequisite tasks in a dictionary.
```
Comment (prachigauriar) [inline Task/Tasks/TSKTask.h]: All the receiver’s.

File: Task/Tasks/TSKTask.h [hunk 3]
```diff
@@ -342,6 +378,29 @@ extern NSString *const TSKTaskDidStartNotification;
...
+ @discussion The returned dictionary maps the prerequisite’s key to its result and will contain
+     NSNull elements for each task that returns nil.
+ @result A dictionary containing the results of all the receiver’s keyed prerequisite tasks.
+ */
+- (NSDictionary *)keyedPrerequisiteResults;
+
+/*!
+ @abstract Returns the result of the receiver’s prerequisite that has the specified key.
```
Comment (prachigauriar) [inline Task/Tasks/TSKTask.h]: Improve this.

File: Task/Workflow/TSKWorkflow.h [hunk 4]
```diff
@@ -219,6 +225,39 @@ extern NSString *const TSKWorkflowTaskKey;
...
+ @discussion The task’s workflow property is set to the receiver, and its prerequisite and keyed 
+     prerequisite tasks are set to the ones specified. Furthermore, for each prerequisite task —
+     keyed or otherwise — the task is added to the prerequisite’s set of dependent tasks. If the 
+     task has any prerequisites, its state is set to pending.
+
+     This is not a thread-safe operation. This method should only execute on one thread at a time.
+ @param task The task to add. May not be nil. May not be a member of any other task workflow.
+ @param prerequisiteTasks The task’s prerequisite tasks. If nil, the task will have no unkeyed
+     prerequisite tasks. Otherwise, each task in the set must have already been added to the receiver.
+ @param keyedPrerequisiteTasks A dictionary that maps the task’s prerequisite names to their
```
Comment (prachigauriar) [inline Task/Workflow/TSKWorkflow.h]: keys not names.

File: Task/Workflow/TSKWorkflow.m [hunk 5]
```diff
@@ -63,11 +63,19 @@ @interface TSKWorkflow ()
...
- @discussion The keys for this map table are TSKTask instances and their values are NSSets.
+ @discussion The keys for this map table are TSKTask instances and their values are NSSets of 
+     TSKTask instances.
  */
 @property (nonatomic, strong, readonly) NSMapTable *prerequisiteTasks;
 
 /*!
+ @abstract A map table that maps a task to its keyed prerequisite tasks.
+ @discussion The keys for this map table are TSKTask instances and their values are NSDictionaries
+     whose keys are strings and whose values are TSKTask instances.
```
Comment (prachigauriar) [inline Task/Workflow/TSKWorkflow.m]: objects that conform to NSCopying

File: Task/Tasks/TSKTask.h (hunk 1 above)
Comment (prachigauriar) [inline Task/Tasks/TSKTask.h]: :ok:

File: Task/Tasks/TSKTask.h (hunk 2 above)
Comment (prachigauriar) [inline Task/Tasks/TSKTask.h]: :ok:

File: Task/Tasks/TSKTask.h (hunk 3 above)
Comment (prachigauriar) [inline Task/Tasks/TSKTask.h]: :ok:

File: Task/Workflow/TSKWorkflow.h (hunk 4 above)
Comment (prachigauriar) [inline Task/Workflow/TSKWorkflow.h]: :ok:

File: Task/Workflow/TSKWorkflow.m (hunk 5 above)
Comment (prachigauriar) [inline Task/Workflow/TSKWorkflow.m]: :ok:

**Comment by prachigauriar** [comment]
@mikehaytm @jnjosh Please review.

File: README.md [hunk 6]
```diff
@@ -8,41 +8,17 @@ makes it easy to cancel tasks, retry failed tasks, and re-run previously complet
...
+
+Task 1.1 adds keyed prerequisites so that a task can retrieve its prerequisite’s results using 
+unique identifiers. This simplifies cases where a task has multiple prerequisites that it needs
+to treat specially. Keyed prerequisites can be specified when adding a task to a workflow using
+`‑[TSKWorkflow addTask:keyedPrerequisiteTasks:]` or 
+`‑[TSKWorkflow addTask:prerequisiteTasks:keyedPrerequisiteTasks:]`. Results from keyed prerequisites
+can be easily retrieved using `‑[TSKTask prerequisiteResultForKey:]`.
+
+Task 1.1 also adds new methods to `TSKTask` so that subclasses can more easily respond to changes in
+a task’s state. See the `SubclassInterface` category of `TSKTask` for more information.
```
Comment (prachigauriar) [inline README.md]: I’m going to glitz this up a bit and add some code samples. It’s a big deal and should read that way.

File: Task/Tasks/TSKBlockTask.m [hunk 7]
```diff
@@ -27,27 +27,42 @@
...
 
 
+@interface TSKBlockTask ()
+
+@property (nonatomic, copy, readonly) NSSet *requiredPrerequisiteKeys;
+
+@end
+
+
+
```
Comment (prachigauriar) [inline Task/Tasks/TSKBlockTask.m]: Remove one line.

File: Task/Tasks/TSKTask.h [hunk 8]
```diff
@@ -243,6 +275,16 @@ extern NSString *const TSKTaskDidStartNotification;
 - (instancetype)initWithName:(NSString *)name NS_DESIGNATED_INITIALIZER;
 
 /*!
+ @abstract Returns the prerequisite keys that the receiver requires to run.
+ @discussion Returns an empty set by default. Subclasses can override this method to return a set of
+     keys for keyed prerequisites that the task requires. When the task is added to a workflow, the
+     workflow will ensure that the task’s required keyed prerequisites are fulfilled. If this property
+     is nil, no prerequisite keys are required.
```
Comment (prachigauriar) [inline Task/Tasks/TSKTask.h]: Will update to mention that subclasses should take care to include their superclass’s keys required keys.

File: Task/Tasks/TSKTask.h [hunk 9]
```diff
@@ -342,6 +387,30 @@ extern NSString *const TSKTaskDidStartNotification;
...
+ @discussion The returned dictionary maps the prerequisite’s key to its result and will contain
+     NSNull elements for each task that returns nil.
+ @result A dictionary containing the results of all the receiver’s keyed prerequisite tasks.
+ */
+- (NSDictionary *)keyedPrerequisiteResults;
+
+/*!
+ @abstract Returns the result of the prerequisite in the receiver’s keyed prerequisites that has
+     the specified key.
+ @result The result for the receiver’s prerequisite task with the specified key. Returns nil if
```
Comment (prachigauriar) [inline Task/Tasks/TSKTask.h]: Add documentation for parameter.

File: Task/Workflow/TSKWorkflow.h [hunk 10]
```diff
@@ -219,6 +225,39 @@ extern NSString *const TSKWorkflowTaskKey;
...
+ @discussion The task’s workflow property is set to the receiver, and its keyed prerequisite tasks are
+     set to the ones specified. Furthermore, the task is added to each of the keyed prerequisite tasks’ 
+     sets of dependent tasks. If the task has any prerequisites, its state is set to pending.
+
+     This is not a thread-safe operation. This method should only execute on one thread at a time.
+ @param task The task to add. May not be nil. May not be a member of any other task workflow.
+ @param keyedPrerequisiteTasks A dictionary that maps the task’s prerequisite names to their 
+     corresponding task. If nil, the task will have no keyed prerequisite tasks. Otherwise, each task
+     in the dictionary must have already been added to the receiver.
+ */
```
Comment (prachigauriar) [inline Task/Workflow/TSKWorkflow.h]: Will likely update the discussion to explain that it’s equivalent to invoking `addTask:prerequisiteTasks:keyedPrerequisiteTasks:`.

**Comment by mikehaytm** [comment]
:+1:

**Pull request merged**